File: verifiers/typescript/src/tiers/tier0Mutation.ts
// tier0Mutation.ts — Stryker driver.
//
// Pins (May 2026): @stryker-mutator/core 9.6.1, @stryker-mutator/vitest-runner 9.6.1.
// Stryker emits mutation reports in the mutation-testing-elements v3.7.x JSON
// schema (https://github.com/stryker-mutator/mutation-testing-elements). We
// shell out to stryker because the public API surface for embedding a
// stryker run is not stable across 9.x. The CLI is invoked with the
// generated `stryker.conf.json` we wrote into the workspace.
//
// Crucible mandate: mutation MUST be diff-scoped. We write the `mutate`
// glob list from the diff's TS/JS files only. The threshold (85% per
// docs/01-architecture/verifier-pipeline.md) is enforced AFTER stryker
// emits its report so we never accept a stryker pass that came from
// running on the whole repo.

import { spawn } from "node:child_process";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";

import { auditDiffPaths } from "../audit.js";
import { classifyDiff } from "../diff.js";
import {
  finalizeTestReport,
  newTestReport,
  type Finding,
  type MutationStats,
  type SurvivedMutant,
  type TestReport,
  type VerificationRequest,
} from "../schema.js";

const MUTATION_THRESHOLD = 0.85; // matches verifier-pipeline.md
const REPORT_PATH = "reports/mutation/mutation.json";

export interface Tier0Options {
  /** Working directory where stryker should run. Defaults to cwd. */
  cwd?: string;
  /** Stryker binary name on PATH. Defaults to "stryker". */
  strykerBin?: string;
  /** Override config-file write path. Defaults to `${cwd}/stryker.conf.json`. */
  configPath?: string;
  /** Wall-clock budget seconds (Crucible default: 30s, max 2 min). */
  wallClockSeconds?: number;
  /**
   * If true, skip actually invoking stryker and emit `tool_unavailable`.
   * The CLI sets this when the binary is missing or in dry-run mode.
   */
  dryRun?: boolean;
  /**
   * For tests: pre-supplied JSON content that bypasses the spawn. When
   * set, we skip the subprocess and parse this string instead.
   */
  preSuppliedReport?: string;
}

// ---- mutation-testing-elements v3.7.x schema shapes (subset) ------------

interface MTEMutant {
  id: string;
  mutatorName: string;
  status: string; // "Killed" | "Survived" | "NoCoverage" | "Timeout" | "CompileError" | "RuntimeError"
  location: {
    start: { line: number; column: number };
    end: { line: number; column: number };
  };
  replacement?: string;
  original?: string;
}

interface MTEFileResult {
  language: string;
  mutants: MTEMutant[];
  source?: string;
}

interface MTEReport {
  schemaVersion: string;
  files: Record<string, MTEFileResult>;
  thresholds?: { high: number; low: number };
  projectRoot?: string;
}

// ---- Stats computation --------------------------------------------------

function classifyMutants(report: MTEReport): {
  killed: number;
  survived: number;
  notCovered: number;
  timeout: number;
  total: number;
  survivors: SurvivedMutant[];
} {
  let killed = 0;
  let survived = 0;
  let notCovered = 0;
  let timeout = 0;
  let total = 0;
  const survivors: SurvivedMutant[] = [];

  for (const [file, fileResult] of Object.entries(report.files)) {
    for (const m of fileResult.mutants) {
      total++;
      switch (m.status) {
        case "Killed":
          killed++;
          break;
        case "Survived":
          survived++;
          survivors.push({
            file,
            line: m.location.start.line,
            mutator: m.mutatorName,
            ...(m.original !== undefined ? { original: m.original } : {}),
            ...(m.replacement !== undefined
              ? { replacement: m.replacement }
              : {}),
          });
          break;
        case "NoCoverage":
          notCovered++;
          break;
        case "Timeout":
          timeout++;
          break;
        // CompileError / RuntimeError are not counted as killed or
        // survived per mutation-testing-elements semantics; ignore.
        default:
          break;
      }
    }
  }

  return { killed, survived, notCovered, timeout, total, survivors };
}

// ---- Config generator ---------------------------------------------------

interface StrykerConfig {
  $schema?: string;
  packageManager?: string;
  reporters: string[];
  mutate: string[];
  testRunner: string;
  coverageAnalysis: string;
  incremental: boolean;
  incrementalFile?: string;
  thresholds?: { high: number; low: number; break: number };
  jsonReporter?: { fileName: string };
  htmlReporter?: { fileName: string };
}

function makeStrykerConfig(mutateFiles: readonly string[]): StrykerConfig {
  return {
    $schema:
      "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
    packageManager: "pnpm",
    reporters: ["json", "clear-text"],
    mutate: mutateFiles.length > 0 ? Array.from(mutateFiles) : ["src/**/*.ts"],
    testRunner: "vitest",
    coverageAnalysis: "perTest",
    incremental: true,
    incrementalFile: ".stryker-tmp/incremental.json",
    thresholds: {
      high: 90,
      low: Math.floor(MUTATION_THRESHOLD * 100),
      break: Math.floor(MUTATION_THRESHOLD * 100),
    },
    jsonReporter: { fileName: REPORT_PATH },
  };
}

// ---- Subprocess driver --------------------------------------------------

function runStryker(
  bin: string,
  cwd: string,
  configPath: string,
  wallClockSeconds: number,
): Promise<{ exitCode: number; stderr: string }> {
  return new Promise((resolve, reject) => {
    const child = spawn(bin, ["run", configPath], {
      cwd,
      stdio: ["ignore", "pipe", "pipe"],
      env: { ...process.env, FORCE_COLOR: "0" },
    });
    let stderr = "";
    let killed = false;
    const timer = setTimeout(
      () => {
        killed = true;
        child.kill("SIGKILL");
      },
      Math.max(1, wallClockSeconds) * 1000,
    );
    child.stderr?.on("data", (b: Buffer) => {
      stderr += b.toString("utf8");
      process.stderr.write(b);
    });
    child.stdout?.on("data", (b: Buffer) => {
      // Stryker logs to stdout by default — forward to stderr so it
      // doesn't pollute the TestReport channel.
      process.stderr.write(b);
    });
    child.on("error", (err) => {
      clearTimeout(timer);
      reject(err);
    });
    child.on("exit", (code) => {
      clearTimeout(timer);
      if (killed) {
        resolve({ exitCode: 124, stderr: stderr + "\n[crucible: killed by wall-clock budget]" });
        return;
      }
      resolve({ exitCode: code ?? 1, stderr });
    });
  });
}

// ---- Public entrypoint --------------------------------------------------

export async function runTier0Mutation(
  req: VerificationRequest,
  opts: Tier0Options = {},
): Promise<TestReport> {
  const startedAt = new Date();
  const cls = classifyDiff(req.diff);
  // Path-pattern guard on the diff — fast-fail if the agent smuggled a
  // reasoning blob into the file path.
  auditDiffPaths(cls.source.map((f) => f.path));

  const wallClockSeconds =
    opts.wallClockSeconds ??
    Math.max(30, Math.min(120, req.budget.wall_clock_cap_seconds || 30));

  const baseReport = newTestReport({
    task_id: req.task_id,
    diff_hash: req.diff.base_sha ?? req.base_sha,
    tier: "tier_0_mutation",
    framework: "stryker-9.6.1",
    reporter_id: "crucible-verify-typescript",
    reporter_version: "2026.06.0-phase4",
    wall_clock_budget_seconds: wallClockSeconds,
    started_at: startedAt,
  });

  if (cls.production.length === 0) {
    return finalizeTestReport(
      {
        ...baseReport,
        verdict: "skipped",
        passed: true,
        error: "no TS/JS production files in diff — Tier 0 skipped",
      },
      new Date(),
    );
  }

  const cwd = opts.cwd ?? process.cwd();
  const configPath = opts.configPath ?? path.join(cwd, "stryker.conf.json");
  const config = makeStrykerConfig(cls.production.map((f) => f.path));

  let strykerStderr = "";
  let preSupplied = opts.preSuppliedReport;

  if (preSupplied === undefined) {
    if (opts.dryRun === true) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "tool_unavailable",
          passed: false,
          framework: "stryker-9.6.1",
          error: "stryker not invoked (dry-run)",
        },
        new Date(),
      );
    }

    try {
      await writeFile(configPath, JSON.stringify(config, null, 2), "utf8");
    } catch (err) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "tool_unavailable",
          passed: false,
          error: `cannot write stryker.conf.json: ${(err as Error).message}`,
        },
        new Date(),
      );
    }

    try {
      const result = await runStryker(
        opts.strykerBin ?? "stryker",
        cwd,
        configPath,
        wallClockSeconds,
      );
      strykerStderr = result.stderr;
      // Stryker exits non-zero when the break threshold is missed — we
      // still parse the report (the threshold check is our own).
    } catch (err) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "tool_unavailable",
          passed: false,
          error: `stryker spawn failed: ${(err as Error).message}`,
        },
        new Date(),
      );
    }
  }

  // Parse the report.
  let raw: string;
  if (preSupplied !== undefined) {
    raw = preSupplied;
  } else {
    const reportFile = path.join(cwd, REPORT_PATH);
    try {
      raw = await readFile(reportFile, "utf8");
    } catch (err) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "failed",
          passed: false,
          error: `cannot read mutation report at ${REPORT_PATH}: ${(err as Error).message}\nstryker stderr: ${strykerStderr.slice(-1000)}`,
        },
        new Date(),
      );
    }
  }

  let parsed: MTEReport;
  try {
    parsed = JSON.parse(raw) as MTEReport;
  } catch (err) {
    return finalizeTestReport(
      {
        ...baseReport,
        verdict: "failed",
        passed: false,
        error: `mutation.json not valid JSON: ${(err as Error).message}`,
      },
      new Date(),
    );
  }

  const stats = classifyMutants(parsed);
  const denom = stats.killed + stats.survived;
  const score = denom > 0 ? stats.killed / denom : 0;
  const passed = score >= MUTATION_THRESHOLD && stats.total > 0;

  const mutation: MutationStats = {
    killed: stats.killed,
    survived: stats.survived,
    total: stats.total,
    score,
    threshold: MUTATION_THRESHOLD,
    diff_scoped: true,
    mutated_files: Object.keys(parsed.files),
    ...(stats.notCovered > 0 ? { not_covered: stats.notCovered } : {}),
    ...(stats.timeout > 0 ? { timeout: stats.timeout } : {}),
    ...(stats.survivors.length > 0
      ? { survived_summary: stats.survivors }
      : {}),
  };

  const findings: Finding[] = stats.survivors.slice(0, 50).map((s) => ({
    category: "mutation_survived",
    severity: "error" as const,
    file: s.file,
    line: s.line,
    detail: `${s.mutator} mutant survived${s.replacement ? ` (replacement: ${s.replacement})` : ""}`,
  }));

  return finalizeTestReport(
    {
      ...baseReport,
      mutation,
      findings,
      verdict: passed ? "passed" : "failed",
      passed,
    },
    new Date(),
  );
}

// ---- Test-only exports --------------------------------------------------

export const __testing__ = {
  classifyMutants,
  makeStrykerConfig,
  MUTATION_THRESHOLD,
};

File: verifiers/rust/src/tiers/tier0_mutation.rs
//! Tier-0 mutation testing via `cargo-mutants` (pinned 27.0.0).
//!
//! Pipeline:
//!
//! 1. Materialise the request's unified diff to a tempfile.
//! 2. Run `cargo mutants --in-diff <tmp>.diff --json --output mutants.out/`.
//! 3. Parse `mutants.out/outcomes.json` (cargo-mutants 27.x schema).
//! 4. Fold the parsed outcomes into `MutationStats`.
//!
//! The Crucible threshold for Rust is 0.85 (85% mutants killed) per the
//! verifier-pipeline doc. `diff_scoped` is **always** `true` (we only
//! ever pass `--in-diff`).

use std::path::Path;
use std::process::Command;
use std::time::Instant;

use serde::Deserialize;

use crate::diff;
use crate::schema::{
    Finding, MutationStats, SurvivedMutant, TestReport, Verdict, VerificationRequest,
};
use crate::tiers::locate;

/// Crucible-wide threshold for Rust mutation score (diff-scoped).
pub const RUST_MUTATION_THRESHOLD: f64 = 0.85;

/// Entrypoint invoked by `tiers::dispatch`.
pub fn run(req: &VerificationRequest, mut report: TestReport) -> TestReport {
    report.framework = "cargo-mutants".to_string();
    let started = Instant::now();

    // Always emit a diff-scoped report skeleton so the dispatcher's
    // validator never trips on `diff_scoped=false`, even when we end
    // up bailing on tool_unavailable below.
    let mut stats = MutationStats {
        threshold: RUST_MUTATION_THRESHOLD,
        diff_scoped: true,
        mutated_files: diff::rust_paths(&req.diff),
        ..MutationStats::default()
    };

    let Some(cargo) = locate("cargo") else {
        report.mutation = Some(stats);
        report.stamp_finished();
        return report.tool_unavailable("cargo not on PATH");
    };

    if locate("cargo-mutants").is_none() {
        // cargo-mutants might still be reachable via `cargo mutants`
        // if a shim exists; fall through and let the cargo invocation
        // surface the real error.
        eprintln!("crucible-verify-rust: warning — cargo-mutants binary not found on PATH; will attempt `cargo mutants` anyway");
    }

    // Materialise the unified diff.
    let diff_body = diff::build_unified_diff(&req.diff);
    if diff_body.trim().is_empty() {
        report.verdict = Verdict::Skipped;
        report.passed = true;
        report.mutation = Some(stats);
        report.stamp_finished();
        report.error = "no rust files in diff — mutation testing skipped".to_string();
        return report;
    }

    let tmp = match tempfile::tempdir() {
        Ok(t) => t,
        Err(e) => {
            report.mutation = Some(stats);
            report.stamp_finished();
            return report.tool_unavailable(format!("tempdir failed: {e}"));
        }
    };
    let diff_path = tmp.path().join("crucible.diff");
    if let Err(e) = std::fs::write(&diff_path, &diff_body) {
        report.mutation = Some(stats);
        report.stamp_finished();
        return report.tool_unavailable(format!("write diff: {e}"));
    }
    let out_dir = tmp.path().join("mutants.out");

    // cargo-mutants runs against the working tree of the executor's
    // repo, which the dispatcher has already mounted into the verifier
    // sandbox at cwd. We DON'T set `current_dir(...)` — that would
    // break cargo's manifest discovery.
    let status = Command::new(&cargo)
        .arg("mutants")
        .arg("--in-diff")
        .arg(&diff_path)
        .arg("--json")
        .arg("--output")
        .arg(&out_dir)
        .stderr(std::process::Stdio::inherit())
        .stdout(std::process::Stdio::inherit())
        .status();

    let exit_ok = match status {
        Ok(s) => {
            // cargo-mutants returns:
            //   0 → all mutants killed
            //   1 → some mutants survived (still a valid run)
            //   2+ → tool failure
            s.code().is_some_and(|c| c < 2)
        }
        Err(e) => {
            report.mutation = Some(stats);
            report.stamp_finished();
            return report.tool_unavailable(format!("cargo mutants spawn: {e}"));
        }
    };

    let outcomes_path = out_dir.join("outcomes.json");
    if !outcomes_path.exists() || !exit_ok {
        report.stamp_finished();
        report.mutation = Some(stats);
        return report.tool_unavailable("cargo-mutants produced no outcomes.json (tool failure)");
    }

    match parse_outcomes(&outcomes_path) {
        Ok(parsed) => {
            stats.killed = parsed.killed;
            stats.survived = parsed.survived;
            stats.not_covered = parsed.not_covered;
            stats.timeout = parsed.timeout;
            stats.total = parsed.total;
            stats.survived_summary = parsed.survived_summary.clone();
            stats.score = if stats.killed + stats.survived == 0 {
                0.0
            } else {
                f64::from(stats.killed) / f64::from(stats.killed + stats.survived)
            };
            let passed = stats.score >= stats.threshold && stats.survived == 0;
            report.verdict = if passed {
                Verdict::Passed
            } else {
                Verdict::Failed
            };
            report.passed = passed;
            report.findings = parsed
                .survived_summary
                .into_iter()
                .map(|m| Finding {
                    category: "mutation_survived".to_string(),
                    severity: "error".to_string(),
                    file: m.file,
                    line: m.line,
                    detail: format!(
                        "{} survived (replaced {:?} with {:?})",
                        m.mutator, m.original, m.replacement
                    ),
                    suggested_fix: String::new(),
                })
                .collect();
        }
        Err(e) => {
            report.error = format!("parse outcomes.json: {e}");
            report.verdict = Verdict::Failed;
            report.passed = false;
        }
    }

    report.mutation = Some(stats);
    report.duration_seconds = started.elapsed().as_secs_f64();
    report.stamp_finished();
    report
}

/// Internal aggregate returned by [`parse_outcomes`].
struct ParsedOutcomes {
    killed: u32,
    survived: u32,
    not_covered: u32,
    timeout: u32,
    total: u32,
    survived_summary: Vec<SurvivedMutant>,
}

/// cargo-mutants 27.x `outcomes.json` shape (relevant subset).
#[derive(Debug, Deserialize)]
struct OutcomesDoc {
    #[serde(default)]
    outcomes: Vec<Outcome>,
}

#[derive(Debug, Deserialize)]
struct Outcome {
    #[serde(default)]
    summary: Option<String>,
    #[serde(default)]
    scenario: Option<Scenario>,
}

#[derive(Debug, Deserialize)]
struct Scenario {
    #[serde(default)]
    mutant: Option<Mutant>,
}

#[derive(Debug, Deserialize)]
struct Mutant {
    #[serde(default)]
    file: String,
    #[serde(default)]
    line: u32,
    #[serde(default)]
    function: String,
    #[serde(default)]
    genre: String,
    #[serde(default)]
    replacement: String,
    #[serde(default)]
    original: String,
}

fn parse_outcomes(path: &Path) -> anyhow::Result<ParsedOutcomes> {
    let raw = std::fs::read_to_string(path)?;
    let doc: OutcomesDoc = serde_json::from_str(&raw)?;
    let mut killed = 0u32;
    let mut survived = 0u32;
    let mut not_covered = 0u32;
    let mut timeout = 0u32;
    let mut survived_summary = Vec::new();
    for o in &doc.outcomes {
        let summary = o.summary.as_deref().unwrap_or("");
        match summary {
            "CaughtMutant" | "Caught" => killed += 1,
            "MissedMutant" | "Missed" => {
                survived += 1;
                if let Some(m) = o.scenario.as_ref().and_then(|s| s.mutant.as_ref()) {
                    survived_summary.push(SurvivedMutant {
                        file: m.file.clone(),
                        line: m.line,
                        mutator: if m.genre.is_empty() {
                            m.function.clone()
                        } else {
                            m.genre.clone()
                        },
                        original: m.original.clone(),
                        replacement: m.replacement.clone(),
                    });
                }
            }
            "Unviable" => not_covered += 1,
            "Timeout" => timeout += 1,
            _ => {}
        }
    }
    let total = killed + survived + not_covered + timeout;
    Ok(ParsedOutcomes {
        killed,
        survived,
        not_covered,
        timeout,
        total,
        survived_summary,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    #[test]
    fn parses_minimal_outcomes() {
        let body = serde_json::json!({
            "outcomes": [
                { "summary": "Caught", "scenario": { "mutant": { "file": "src/a.rs", "line": 12, "function": "f", "genre": "BinaryOperator", "original": "+", "replacement": "-" } } },
                { "summary": "Missed", "scenario": { "mutant": { "file": "src/b.rs", "line": 7, "function": "g", "genre": "BooleanLiteral", "original": "true", "replacement": "false" } } },
                { "summary": "Unviable" },
                { "summary": "Timeout" }
            ]
        });
        let mut f = tempfile::NamedTempFile::new().unwrap();
        write!(f, "{}", body).unwrap();
        let parsed = parse_outcomes(f.path()).unwrap();
        assert_eq!(parsed.killed, 1);
        assert_eq!(parsed.survived, 1);
        assert_eq!(parsed.not_covered, 1);
        assert_eq!(parsed.timeout, 1);
        assert_eq!(parsed.total, 4);
        assert_eq!(parsed.survived_summary.len(), 1);
        assert_eq!(parsed.survived_summary[0].file, "src/b.rs");
        assert_eq!(parsed.survived_summary[0].mutator, "BooleanLiteral");
    }
}
